### DISCUSSION TITLE
Explicitly load a relationship on an instance

### DISCUSSION THREAD

**Discussion by ay-ramp** [sqlalchemy/sqlalchemy#10606] [Usage Questions]
Hi,

If I have an instance of model like:

```
class MyModel:
   ...

   related = relationship("MyOtherModel", lazy="raise")


my_obj = db.session.query(select(MyModel).where(MyModel.id==123)).scalar_one()
```

is there any way that i can override the `lazy="raise"` _on that loaded instance_ to explicitly fetch the related item? Something like:

```
with allow_lazy_loading():
    related = my_obj.related
```

or

```
preload(my_obj, 'related')
```
?

I would prefer not to use `lazy="select"` here, because it's too easy to introduce unintentional data-loads in code where I'd really like it to happen on the initial query. In this case, I also easily ensure that the right load options are present at query time.

**Comment by CaselIT** [comment]
Hi,

I think you can use session.get passing the options accordingly. A better option would be to change the query so that it includes the eager loader when you need it

**Reply by ay-ramp** [reply]
> A better option would be to change the query so that it includes the eager loader when you need it

mind elaborting on this one? if it requires that i know which related options i'll need at first query time, unfortunately won't work for me in this case (lots of nested functions / branches during a mutation that i can't reasonably specify static dependencies for)

> I think you can use session.get passing the options accordingly. 

i tried this and am not seeing it work, any pointers?

```py
from typing import Any

from sqlalchemy import Column, Integer, create_engine, select, event
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import (
    Session,
    with_loader_criteria,
    relationship,
    joinedload,
    selectinload,
)

Base: Any = declarative_base()


class B(Base):
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)
    a_id = Column(Integer)

    a = relationship("A", primaryjoin="foreign(B.a_id) == A.id", back_populates="b", lazy="raise")


class A(Base):
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)

    b = relationship("B", back_populates="a", primaryjoin="foreign(B.a_id) == A.id", lazy="raise")



e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

session = Session(e)


a = A(id=1)
b = B(id=1, a=a)
session.add(a)
session.commit()

other_session = Session(e)

print("Fetching a_1 first time")
a_1 = other_session.get(A, 1)
# a_1.b  # this raises "'A.b' is not available due to lazy='raise'"

print("Fetching a_2 -- this should not emit sql")
a_2 = other_session.get(A, 1)

print("Fetching a_3 with load options -- this will emit sql")
a_3 = other_session.get(A, 1, options=selectinload(A.b))
a_3.b  # this still raises: "'A.b' is not available due to lazy='raise'"
```

with relevant logs:

```
Fetching a_1 first time
2023-11-09 07:41:41,948 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-11-09 07:41:41,949 INFO sqlalchemy.engine.Engine SELECT a.id AS a_id 
FROM a 
WHERE a.id = ?
2023-11-09 07:41:41,949 INFO sqlalchemy.engine.Engine [generated in 0.00005s] (1,)
Fetching a_2 -- this should not emit sql
Fetching a_3 with load options -- this will emit sql
Traceback (most recent call last):
...
sqlalchemy.exc.InvalidRequestError: 'A.b' is not available due to lazy='raise'
```

(doesn't look like it refetches/applies those options)

i'm using 1.4 if it makes a difference (would be good to know if this gets better in 2)

**Reply by CaselIT** [reply]
a_3 should work if you pass `populate_existing=True`

**Reply by CaselIT** [reply]
> mind elaborting on this one? if it requires that i know which related options i'll need at first query time, unfortunately won't work for me in this case (lots of nested functions / branches during a mutation that i can't reasonably specify static dependencies for)

I guess it's not viable then

**Reply by ay-ramp** [reply]
hmm, i see that work but it appears to refetch my model each time:

```
Fetching a_1 first time
2023-11-09 08:00:58,559 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-11-09 08:00:58,560 INFO sqlalchemy.engine.Engine SELECT a.id AS a_id 
FROM a 
WHERE a.id = ?
2023-11-09 08:00:58,560 INFO sqlalchemy.engine.Engine [generated in 0.00007s] (1,)
Fetching a_2 -- this should not emit sql
Fetching a_3 with load options -- this will emit sql
2023-11-09 08:00:58,561 INFO sqlalchemy.engine.Engine SELECT a.id AS a_id 
FROM a 
WHERE a.id = ?
2023-11-09 08:00:58,561 INFO sqlalchemy.engine.Engine [generated in 0.00008s] (1,)
2023-11-09 08:00:58,562 INFO sqlalchemy.engine.Engine SELECT b.a_id AS b_a_id, b.id AS b_id 
FROM b 
WHERE b.a_id IN (?)
2023-11-09 08:00:58,562 INFO sqlalchemy.engine.Engine [generated in 0.00006s] (1,)
Fetching a_4 with load options -- shouldn't emit sql
2023-11-09 08:00:58,562 INFO sqlalchemy.engine.Engine SELECT a.id AS a_id 
FROM a 
WHERE a.id = ?
2023-11-09 08:00:58,562 INFO sqlalchemy.engine.Engine [cached since 0.001362s ago] (1,)
2023-11-09 08:00:58,563 INFO sqlalchemy.engine.Engine SELECT b.a_id AS b_a_id, b.id AS b_id 
FROM b 
WHERE b.a_id IN (?)
2023-11-09 08:00:58,563 INFO sqlalchemy.engine.Engine [cached since 0.0007341s ago] (1,)
```

```
print("Fetching a_1 first time")
a_1 = other_session.get(A, 1)
# a_1.b  # this raises "'A.b' is not available due to lazy='raise'"

print("Fetching a_2 -- this should not emit sql")
a_2 = other_session.get(A, 1)

print("Fetching a_3 with load options -- this will emit sql")
a_3 = other_session.get(A, 1, options=[selectinload(A.b)], populate_existing=True)
a_3.b 

print("Fetching a_4 with load options -- shouldn't emit sql")
a_4 = other_session.get(A, 1, options=[selectinload(A.b)], populate_existing=True)
```

how can i *only* emit sql if `A.b` hasn't yet been populated, while using `lazy="raise"`?

**Reply by ay-ramp** [reply]
and for context, the reason why is that my code is encapsulated in such a way (for better or worse) that actually code doesn't know whether it's been previously fetched or not

**Comment by zzzeek** [comment]
> is there any way that i can override the lazy="raise" on that loaded instance to explicitly fetch the related item? Something like:

only if you query for it again using populate_existing, where the new lazyload option is given.  you can use `get()` for this purpose and here is a demonstration

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import lazyload
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    bs: Mapped[list[B]] = relationship("B", lazy="raise")


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(bs=[B()]))
s.commit()

a1 = s.scalar(select(A))

s.get(A, a1.id, options=[lazyload(A.bs)], populate_existing=True)
a1.bs
```

**Reply by ay-ramp** [reply]
my real use case is something like this:

```
def perform_action(a: A):
   # need to use b to perform action
   # TODO: add some code here that makes this "safe" / won't raise
   bs = a.bs
   print("example action: printing all the bs", bs)

def perform_other_action(a: A):
   # need to use b to perform action
   # TODO: add some code here that makes this "safe" / won't raise
   bs = a.bs
   print("example other action: printing all the bs with a different prefix", bs)
```

and `perform_action` and `perform_other_action` might be called in any order, and don't necessarily have a clean way to explain to code further up the stack that `A` must have `bs` loaded. 

given that context, i think this still doesn't quite work because i'd have to add something like `s.get(A, a1.id, options=[lazyload(A.bs)], populate_existing=True)` at the beginning of each function, which will end up emitting sql to refetch (i think? see https://github.com/sqlalchemy/sqlalchemy/discussions/10606#discussioncomment-7522474).

the [ecto api](https://hexdocs.pm/ecto/Ecto.Repo.html#c:preload/3) is pretty nice on this front with an explicit `loaded_obj |> preload(:relationship)`

**Reply by zzzeek** [reply]
you can make a function that does the "s.get" for you?  is it too verbose?   It can be entirely encapsulated into a function like `set_lazyload(object)`.   the current session is available from `inspect(object).session`, etc.  What's the issue exactly?

**Reply by zzzeek** [reply]
it also runs a query, so that's not ideal either

**Reply by zzzeek** [reply]
overall the best solution here is to organize your code so that the things you want up front are loaded.  What's the point of using `lazy="raise"` in your mapping if downstream functions can simply turn it off anyway?

**Reply by ay-ramp** [reply]
yeah, running the query is the part that's not ideal. basically want something that:
* is explicit that it *might* result in a network fetch (not a property access)
* fetches if the relationship hasn't been fetched for the object in the identity map already
* doesn't fetch if the relationship has been fetched for the object in the identity map already

**Reply by ay-ramp** [reply]
> What's the point of using lazy="raise" in your mapping if downstream functions can simply turn it off anyway?

for the same models, we have two different patterns. most of our traffic on the models is on read endpoints, where we can have very declarative sets of dependencies and lazy="raise" works really well. other traffic is on write endpoints that are organized above and we basically want to be able to get similar behavior to `lazy="select"` (but ideally being a bit more explicit about it)

**Reply by zzzeek** [reply]
OK so are you loading those objects with queries?  (if so then use lazyload() params for that codepath) ?    or are they being persisted with `session.commit()` and the issue is that the attributes expired ? (use expire_on_commit=False)

There are of course ways SQLAlchemy can have APIs to change "raiseload" flags for relationships and deferred columns but we can't add random APIs like this without extremely clear ideas.  There are APIs that could load the attribute immediately, there are APIs that could disable *all* raiseloads, there are APIs that could try to do what you mention which is switch off the "raise" flag for specific attributes withotu actually loading them, however that last one would be the most intricate feature add and would require a lot of changes, and I will note to myself and others this API *must* work for deferred columns too so we dont want to forget that.   I dont want to just jump into that without a very clear idea of how you are in this situation in the first place

**Reply by zzzeek** [reply]
also we are past the window for big feature adds in 2.0 so this would not be in 2.0 in any case.    I can always write for you an internals hack that makes this "happen" but I can't guarantee it wont break on new versions, that's the best I can do here.

**Reply by ay-ramp** [reply]
> OK so are you loading those objects with queries? (if so then use lazyload() params for that codepath) ?

this is our situation -- probably something like this is the best we can do for now. appreciate the offer of the internals hack but we'd probably prefer to rely on it.

thank you for talking this through!

**Reply by CaselIT** [reply]
maybe you can manually run a query and use `set_committed_value`
For example, assuming the relationship is of models `M`
```py
def load_thing(inst, thing):
    is_ = sa.inspect(inst)
    if thing in is_.unloaded:
        print("load", thing)
        p = is_.session.scalars(sa.select(M).where(M.xid == inst.id)).all()
        set_committed_value(inst, thing, p)
    else:
        print("loaded")
```

**Comment by CaselIT** [comment]
@zzzeek any reason by doing `session.refresh(a1, ['bs'])` does also a query on a1? Otherwise this could the a simple solution for this use case

From the example above https://github.com/sqlalchemy/sqlalchemy/discussions/10606#discussioncomment-7522592 add the likes
```py
s.refresh(a1,[ 'bs'])
print(a1.bs)
```
these log
```
2023-11-09 19:29:41,979 INFO sqlalchemy.engine.Engine SELECT a.id
FROM a
WHERE a.id = ?
2023-11-09 19:29:41,979 INFO sqlalchemy.engine.Engine [generated in 0.00013s] (1,)
2023-11-09 19:29:41,980 INFO sqlalchemy.engine.Engine SELECT b.id AS b_id, b.a_id AS b_a_id
FROM b
WHERE ? = b.a_id
2023-11-09 19:29:41,980 INFO sqlalchemy.engine.Engine [generated in 0.00013s] (1,)
[<__main__.B object at 0x0000015C12C010A0>]
```

the first query seems totally unnecessary here

**Reply by zzzeek** [reply]
yes it has to do the "id" refresh, there was a whole issue involving this.

 i mean in theory, maybe it doesnt, but there's no short path right now

**Reply by CaselIT** [reply]
Do you remember which one was the issue?

**Reply by zzzeek** [reply]
from the changelog, there's three issues together so one of these: #8703, #8997, #8996

**Reply by zzzeek** [reply]
and then specifically my comment here: https://github.com/sqlalchemy/sqlalchemy/issues/8701#issuecomment-1288200238

**Reply by CaselIT** [reply]
Understood. Adding a flag to enable it is not the best option, since I can't come up with a name that would mean anything to users

It may make sense adding something similar to this feature https://hexdocs.pm/ecto/Ecto.Repo.html#c:preload/3 

> Preloads all associations on the given struct or structs.

> This is similar to [Ecto.Query.preload/3](https://hexdocs.pm/ecto/Ecto.Query.html#preload/3) except it allows you to preload structs after they have been fetched from the database.

> In case the association was already loaded, preload won't attempt to reload it.

translated to python this would be something like
```py
session.load_relationships(instance_or_list, options, ...)
```

or we could even restrict this to selectin

```py
session.selectin_relationships(instance_or_list, relationships, ...)
```

**Reply by zzzeek** [reply]
I think refresh() should do the thing we want and not load the lead table. it's just a big code change that's all

**Reply by CaselIT** [reply]
> I think refresh() should do the thing we want and not load the lead table. it's just a big code change that's all

yes, that's probably the thing that does most sense.
shall I open an issue so that we track that feature?

**Reply by zzzeek** [reply]
maybe.  I really would need to look and see what this change entails.    im not fully sure what my previous thinking was.   sure open an issue for 2.1

**Reply by zzzeek** [reply]
here's the first question(s) to be answered.

Q: What happens if the relationship A.bs is mapped as lazy='joined', and we say `s.refresh(a1, ['bs'])` ?   

Q: What happens if someone does this:   `s.refresh(a1, ['bs'], options=[joinedload(A.bs)])` ?  or this:  `s.refresh(a1, ['bs'], options=[selectinload(A.bs)])`  ?   See this is a big problem, if i honor that then I have to go through the whole process of compiling a select against "a1" first, or we have to build a whole system to run through ORM options without a query.   If we just make it that `s.refresh(obj, ..., options=[anything])`, that means you get a query against a1, that would solve that , but this is not intuitive

**Reply by CaselIT** [reply]
> Q: What happens if the relationship A.bs is mapped as lazy='joined', and we say s.refresh(a1, ['bs']) ?

Refresh could just refresh the relationship without implying any loading style. It could continue to use the the simplified version selectin that uses `=` that's using now

> Q: What happens if someone does this: s.refresh(a1, ['bs'], options=[joinedload(A.bs)]) ? or this: s.refresh(a1, ['bs'], options=[selectinload(A.bs)]) ? See this is a big problem, if i honor that then I have to go through the whole process of compiling a select against "a1" first, or we have to build a whole system to run through ORM options without a query. If we just make it that s.refresh(obj, ..., options=[anything]), that means you get a query against a1, that would solve that , but this is not intuitive

refresh would have no options. it can refresh a relationship but gives no control to the user on how it does that.

If the question was more "what to do if the user passes `selectinload(A.bs).selectinload(B.cs)` I would say that if refresh has no options then this use case would not be supported

**Reply by zzzeek** [reply]
oh hooray, refresh doesnt take options right now!   I was thinking of session.get().   great

**Reply by CaselIT** [reply]
added https://github.com/sqlalchemy/sqlalchemy/issues/10617

**Reply by CaselIT** [reply]
> oh hooray, refresh doesnt take options right now! I was thinking of session.get(). great

indeed, so it can just do the simplest thing it already does, just skipping the model table query
